contrib/scripts/kind.sh: fix DNS resolution on nodes

atykhyy opened the pull request:
This PR fixes contrib/scripts/kind.sh so that connectivity test for #34024 (Enable L7 DNS proxy for nodes) works.

The test fails in the default cluster created by `contrib/scripts/kind.sh` because, as has been described in #23330, #32489 and many other issues in between, dockerd is routing packets to its internal DNS server with a netfilter rule. This rule snatches packets off the datapath before they arrive at DNS proxy. #24713 worked around this for pods in clusters created by `contrib/scripts/kind.sh` by replacing the nameserver in coredns pods' /etc/resolv.conf. Unfortunately this approach cannot be made to work reliably in the host: kubelets need to be able to resolve node hostnames to find their k8s cluster peers, and so they need to either somehow refer to dockerd's internal DNS server for node hostnames, or to have all node hostnames mapped to addresses in /etc/hosts. (It is possible to overwrite host /etc/resolv.conf to replace dockerd's DNS with 8.8.8.8 several seconds after kubelet has started - after it has resolved node hostnames - but this is not reliable.) 

This PR adds a kubelet drop-in that replaces the nameserver configured by docker with dnsmasq defaulting to 8.8.8.8, but deferring local lookups to docker DNS so that kubelets can resolve nodes by name.

bimmlerd commented on contrib/scripts/kind-setup-dns.sh:
```diff
@@ -0,0 +1,19 @@
+# no shebang, this script is never run directly
```
Nit: Please also leave a brief paragraph explaining why this file exists, ideally with a reference to the github issue (like you did in kind.sh)

atykhyy replied:
Done.

bimmlerd reviewed (commented):
Seems sane, one nit to address please :)

atykhyy committed:
kind.sh: fix DNS resolution on nodes, make external DNS configurable

Adds kubelet drop-in that replaces the nameserver configured by docker
with dnsmasq defaulting to an external DNS server, but deferring local lookups
to docker DNS so that kubelets can resolve nodes by name.
This is required because in case of BPF Host Routing we bypass iptables thus
breaking DNS (#23330), and because those iptables rules prevent L7 DNS proxy
from working.

External DNS server defaults to 1.1.1.1, but is configurable
via kind.sh command line argument or EXTERNAL_DNS environment variable.

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>
